data/libs/Passengers.lua

```diff
@@ -0,0 +1,183 @@
+-- Copyright © 2008-2024 Pioneer Developers. See AUTHORS.txt for details
+-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+local Equipment = require 'Equipment'
+local EquipSet = require 'EquipSet'
+
+local utils = require 'utils'
+
+-- Module: Passengers
+--
+-- Helper utilities for working with passengers onboard a ship
+local Passengers = {}
+
+
+-- Function: CountOccupiedCabins
+--
+-- Return the number of occupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountOccupiedCabins(ship)
+	return ship["cabin_occupied_cap"] or 0
+end
+
+-- Function: CountFreeCabins
+--
+-- Return the number of unoccupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountFreeCabins(ship)
+	return (ship["cabin_cap"] or 0) - (ship["cabin_occupied_cap"] or 0)
+end
+
+-- Function: GetOccupiedCabins
+--
+-- Return a list of currently occupied cabin equipment items
+--
+---@param ship Ship
+---@return Equipment.CabinType[]
+function Passengers.GetOccupiedCabins(ship)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetNumPassengers() > 0
+	end)
+end
+
+-- Function: GetFreeCabins
+--
+-- Return a list of currently free passenger cabins
+--
+---@param ship Ship
+---@param numPassengers integer?
+---@return Equipment.CabinType[]
+function Passengers.GetFreeCabins(ship, numPassengers)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetFreeBerths() >= (numPassengers or 1)
```

Comment:
Nitpick: I would add `CabinType:IsFree()`  to the Cabin class and use that here and in `GetOccupiedCabins()`.

Also, what does `numPassengers` have to do with this? It seems that this function only returns cabins which have at least `numPassengers` free berths, so the functionality does not match the documentation.

Finally, this implies that if I'm trying to add X passengers to my ship, but none of the cabins have enough free room for X, but overall I have more than X free berths in the ship, then I can't split up the group amongst available cabins (which may be a feature or a bug, depending on the use-case..)

Reply:
This is entirely dependent on the calling code setting the `numPassengers` parameter. If the calling code is trying to seat a group that can be split, then it will call `GetFreeCabins(ship, 1)` and iterate through the result, seating passengers into each cabin until all passengers are seated or the cabin is full.

If the calling code is trying to seat a group that cannot be split, then it will call `GetFreeCabins(ship, N)` and complain to the player that their ship doesn't have room for the group if no cabins are returned.

I've updated the documentation to clarify this a bit better.

Note: `GetFreeBerths()` != `not IsEmpty()`. As the only callsite I've found for `GetNumPassengers() > 0` is right above, I've declined to add an `IsEmpty()` method for the reason of lack of demand.